src/pv.rs

```diff
@@ -43,28 +44,118 @@ fn packet_dump(packet: &PvPacket) {
 }
 
 #[derive(Debug)]
-pub struct PvPacket {
+pub struct Packet {
     pub start: u32,       // payload offset pointing the start of payload. ex. 256
     pub end: u32,         // payload offset point the end of payload. ex. 1000
-    pub size: u32,        // total size of buffer. ex. 2048
+    pub buffer_size: u32, // total size of buffer. ex. 2048
     pub buffer: *mut u8,  // buffer address.
     private: *mut c_void, // DO NOT touch this.
 }
 
-impl PvPacket {
-    fn new() -> PvPacket {
-        PvPacket {
+impl Packet {
+    fn new() -> Packet {
+        Packet {
             start: 0,
             end: 0,
-            size: 0,
+            buffer_size: 0,
             buffer: std::ptr::null_mut(),
             private: std::ptr::null_mut(),
         }
     }
+
+    /* check whether the new header can be attached to payload from the start of packet
+    add header to payload and return Ok if it's possible, or return Err */
+    pub fn attach_header(&mut self, header: Vec<u8>) -> Result<(), ()> {
+        if self.start >= header.len() as u32 {
+            // header can be attached in empty space of headroom
+
+            self.start = self.start - header.len() as u32;
+
+            unsafe {
+                copy_nonoverlapping(
+                    header.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    header.len(),
+                );
+            }
+
+            Ok(())
+        } else if header.len() as u32 <= self.buffer_size - self.end + self.start {
+            // header can't be attached in headroom but can be by shifting payload
+            unsafe {
+                // move payload (memmove)
+                copy(
+                    self.buffer.offset(self.start as isize),
+                    self.buffer.offset(header.len() as isize),
+                    (self.end - self.start) as usize,
+                );
+
+                // set start and end of data
+                self.start = 0;
+                self.end = self.end + (header.len() as u32 - self.start);
+
+                // copy header data (memcpy)
+                copy_nonoverlapping(
+                    header.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    header.len(),
+                );
+            }
+
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    // replace payload with new data from *start and make the rest of original payload zero
+    pub fn replace_data_from_start(&mut self, new_data: &Vec<u8>) -> Result<(), ()> {
```

Comment:
Total nonsense.

- Why zero padding on non-used area?
- As it replaces whole data
  - why preserve packet.start?
  - "from_start" is meaningless

Reply:
About padding zeros, I thought the packet should have the only datum not overlapped data if data should be replaced.
But I noticed that it's meaningless for the following reason.
The reason is that packet metadata, `Packet`, has `start` and `end` field, and XSK sends packets with only information of `data length` and `pointer`, so that padding zeros rather makes performance lower.

About preserving `packet.start`, I thought preserving default headroom(256 bytes) until the packet has no space might be good if new headers have to be attached in the future processing procedure. Repeatedly shifting data and attaching headers may lower the performance after headroom is used at the first processing step.
However in the point of view that`packet.start` can move anywhere, the preserving seems meaningless

Reply:
I made the function simple